File: codejam/python/2010/2/2_GrazingGoogleGoats.py
import math
import sys
from collections import deque, namedtuple

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007

Line = namedtuple('Line','m b cx cy r')
Pt = namedtuple('Pt','x y')
def rot2d(x,y,ang) : return (x*math.cos(ang)-y*math.sin(ang),x*math.sin(ang)+y*math.cos(ang))
def invertPoint(x,y) : return (x/(x*x+y*y),y/(x*x+y*y))
def invertCircleToLine(cx,cy) :
    ## We shouldn't have vertical lines after the rotation, so use (m,b) for y=mx+b to rep line
    m = -cx/cy; xi,yi = invertPoint(2*cx,2*cy); b = yi-m*xi; return (m,b)
def intersection(l1,l2) :
    x = (l2.b-l1.b)/(l1.m-l2.m); y = l1.m*x+l1.b; return (x,y)
def circleSegment(x1,y1,x2,y2,r) :
    ang1 = math.atan2(y1,x1)
    ang2 = math.atan2(y2,x2)
    ang = ang2-ang1 if ang2-ang1 >= 0 else 2*math.pi+ang2-ang1
    ans = 0.5*r*r* ( ang - math.sin(ang))
    return ans 
def polyArea(plist) :
    n = len(plist); area = 0.0
    for i in range(n-1) : area += plist[i].x*plist[i+1].y-plist[i].y*plist[i+1].x
    area += plist[n-1].x*plist[0].y-plist[n-1].y*plist[0].x
    return 0.5*abs(area)

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        N,M = gi(),gi(); PX,PY = fill2(N); QX,QY = fill2(M); ansarr = [0.00] * M
        eps = 1e-12
        for idx in range(M) :
            qx,qy = QX[idx],QY[idx]
            angs = [math.atan2(y-qy,x-qx) for (x,y) in zip(PX,PY)]
            angs.sort()
            gaps = [angs[i]-angs[i-1] for i in range(N)]; gaps[0] += 2*math.pi
            bestgap = max(gaps)
            if bestgap <= math.pi + eps : ansarr[idx] = 0; continue
            bestidx = gaps.index(bestgap)
            rotang = 0.5*math.pi-(angs[bestidx]-0.5*bestgap) ## Make center of gap point up
            lines = []
            for i in range(N) :
                rx,ry = rot2d(PX[i]-qx,PY[i]-qy,rotang)
                m,b = invertCircleToLine(rx,ry)
                lines.append(Line(m,b,rx,ry,math.sqrt(rx*rx+ry*ry)))
            lines.sort(); lines.reverse()
            ## Lower envelope routine
            st = []
            for l in lines :
                while len(st) >= 2 :
                    x1,y1 = intersection(st[-2],st[-1])
                    x2,y2 = intersection(st[-1],l)
                    if x2 < x1 : st.pop()
                    else : break
                st.append(l)
            ## Calculate the area
            poly = [Pt(0,0)]; area = 0; lx,ly = 0,0
            for i in range(len(st)-1) :
                cx,cy,r = st[i].cx,st[i].cy,st[i].r
                xx,yy = intersection(st[i],st[i+1])
                x,y = invertPoint(xx,yy)
                poly.append(Pt(x,y))
                adder = circleSegment(lx-cx,ly-cy,x-cx,y-cy,r)
                area += adder
                lx,ly = x,y
            cx,cy,r = st[-1].cx,st[-1].cy,st[-1].r
            adder = circleSegment(lx-cx,ly-cy,0-cx,0-cy,r)
            area += adder
            area += polyArea(poly)
            ansarr[idx] = area
        ansstr = " ".join(["%.17g" % x for x in ansarr])
        print(f"Case #{tt}: {ansstr}")

if __name__ == "__main__" :
    main()

File: codejam/cpp/2010/2/2_GrazingGoogleGoats.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));
vector<vector<ll>> twodi(ll N, ll M, ll v) { vector<vector<ll>> res(N); for (auto &vv : res) vv.resize(M,v); return res; }

struct Line { double m; double b; double cx; double cy; double r; };
struct Pt { double x; double y; };
bool operator<(Line &a, Line &b) { return a.m > b.m; }
Pt rot2d(double x, double y, double ang) { return Pt{x*cos(ang)-y*sin(ang),x*sin(ang)+y*cos(ang)}; }
Pt invertPoint(double x, double y) { return Pt{x/(x*x+y*y),y/(x*x+y*y)}; }
Line invertCircleToLine(double cx, double cy) {
    auto m = -cx/cy; auto p = invertPoint(2.0*cx,2.0*cy); auto b = p.y-m*p.x; return Line{m,b,cx,cy,sqrt(cx*cx+cy*cy)};
}
Pt intersection(Line &l1, Line &l2) { auto x = (l2.b-l1.b)/(l1.m-l2.m); auto y = l1.m*x+l1.b; return Pt{x,y}; }
double circleSegment(double x1, double y1, double x2, double y2, double r) {
    auto ang1 = atan2(y1,x1); auto ang2 = atan2(y2,x2); auto ang = ang2-ang1 >= 0 ? ang2-ang1 : 2.0*PI+ang2-ang1;
    return 0.5*r*r*(ang - sin(ang));
}
double polyArea(vector<Pt> &plist) {
    auto n = plist.size(); auto area = 0.0;
    FOR(i,n-1) area += plist[i].x*plist[i+1].y-plist[i].y*plist[i+1].x;
    area += plist[n-1].x*plist[0].y-plist[n-1].y*plist[0].x;
    area *= 0.5;
    return area;
}
int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll N,M; cin >> N >> M;
        vi PX(N),PY(N); FOR(i,N) cin >> PX[i] >> PY[i];
        vi QX(M),QY(M); FOR(i,M) cin >> QX[i] >> QY[i];
        vector<double> ansarr(M,0.00); double eps = 1e-12;
        FOR(idx,M) {
            auto qx = QX[idx]; auto qy = QY[idx];
            vector<double> angs(N); FOR(i,N) angs[i] = atan2(double(PY[i])-qy,double(PX[i])-qx);
            sort(angs.begin(),angs.end());
            vector<double> gaps(N); FOR(i,N) gaps[i] = angs[i]-angs[(N+i-1)%N]; gaps[0] += 2.0*PI;
            double bestgap(0.0); ll bestidx(0);
            FOR(i,N) if (gaps[i] > bestgap) { bestgap = gaps[i]; bestidx = i; }
            if (bestgap <= PI+eps) { ansarr[idx] = 0.00; continue; }
            double rotang = 0.5*PI-(angs[bestidx]-0.5*bestgap);
            vector<Line> lines;
            FOR(i,N) {
                auto p = rot2d(double(PX[i])-qx,double(PY[i])-qy,rotang);
                lines.push_back(invertCircleToLine(p.x,p.y));
            }
            sort(lines.begin(),lines.end());
            vector<Line> st;
            for (auto &l : lines) {
                while(st.size() >= 2) {
                    auto p1 = intersection(st[st.size()-2],st[st.size()-1]);
                    auto p2 = intersection(st[st.size()-1],l);
                    if (p2.x < p1.x) { st.pop_back(); } else break;
                }
                st.push_back(l);
            }
            vector<Pt> poly;
            poly.push_back(Pt{0.00,0.00}); double area = 0.00; double lx = 0.00; double ly = 0.00;
            auto n = st.size();
            FOR(i,n-1) {
                auto cx = st[i].cx; auto cy = st[i].cy; auto r = st[i].r;
                auto p1 = intersection(st[i],st[i+1]);
                auto p2 = invertPoint(p1.x,p1.y);
                poly.push_back(p2);
                auto adder = circleSegment(lx-cx,ly-cy,p2.x-cx,p2.y-cy,r);
                area += adder;
                lx = p2.x; ly = p2.y;
            }
            auto cx = st[n-1].cx; auto cy = st[n-1].cy; auto r = st[n-1].r;
            auto adder = circleSegment(lx-cx,ly-cy,0.0-cx,0.0-cy,r);
            area += adder;
            area += polyArea(poly);
            ansarr[idx] = area;
        }
        cout << "Case #" << tt << ":";
        FOR(i,M) { cout << ' ' << setprecision(17) << ansarr[i]; }
        cout << '\n';
    }
}
